src/Matrix.cpp

```diff
@@ -0,0 +1,465 @@
+    log_printf(ERROR, "Unable to increment Matrix value for group_to %i"
+               " which is not between 0 and %i", group_to, _num_groups-1);
+  
+  /* Atomically increment the Matrix value from the
+   * temporary array using mutual exclusion locks */
+  omp_set_lock(&_cell_locks[cell_to]);
+
+  int row = cell_to*_num_groups + group_to;
+  int col = cell_from*_num_groups + group_from;
+  _LIL[row][col] += val;
+  
+  /* Release Matrix cell mutual exclusion lock */
+  omp_unset_lock(&_cell_locks[cell_to]);
+
+  /* Set global modified flag to true */
+  _modified = true;
+}
+
+
+/**
+ * @brief Set a value in the matrix.
+ * @detail This method takes a cell and group of origin (cell/group from)
+ *         and cell and group of destination (cell/group to) and floating
+ *         point value. The origin and destination are used to compute the
+ *         row and column in the matrix. The location specified by the
+ *         row/column is set to val.
+ * @param cell_from The origin cell.
+ * @param group_from The origin group.
+ * @param cell_to The destination cell.
+ * @param group_from The destination group.
+ * @param val The value used to set the row/column location.
+ */
+void Matrix::setValue(int cell_from, int group_from,
+                      int cell_to, int group_to, FP_PRECISION val) {
+
+  if (cell_from >= _num_x*_num_y || cell_from < 0)
+    log_printf(ERROR, "Unable to set Matrix value for cell_from %i"
+               " which is not between 0 and %i", cell_from, _num_x*_num_y-1);
+  else if (cell_to >= _num_x*_num_y || cell_to < 0)
+    log_printf(ERROR, "Unable to set Matrix value for cell_to %i"
+               " which is not between 0 and %i", cell_from, _num_x*_num_y-1);
+  else if (group_from >= _num_groups || group_from < 0)
+    log_printf(ERROR, "Unable to set Matrix value for group_from %i"
+               " which is not between 0 and %i", group_from, _num_groups-1);
+  else if (group_to >= _num_groups || group_to < 0)
+    log_printf(ERROR, "Unable to set Matrix value for group_to %i"
+               " which is not between 0 and %i", group_to, _num_groups-1);
+  
+  /* Atomically set the Matrix value from the
+   * temporary array using mutual exclusion locks */
+  omp_set_lock(&_cell_locks[cell_to]);
+
+  int row = cell_to*_num_groups + group_to;
+  int col = cell_from*_num_groups + group_from;
+  _LIL[row][col] = val;
+  
+  /* Release Matrix cell mutual exclusion lock */
+  omp_unset_lock(&_cell_locks[cell_to]);
+
+  /* Set global modified flag to true */
+  _modified = true;  
+}
+
+
+/**
+ * @brief Clear all values in the matrix list of lists.
+ */
+void Matrix::clear() {
+  for (int i=0; i < _num_rows; i++)
+    _LIL[i].clear();
+
+  _modified = true;
+}
+
+
+/**
+ * @brief Convert the matrix lists of lists to compressed row (CSR) storage 
+ *        form.
+ */
+void Matrix::convertToCSR() {
+  
+  /* Get number of nonzero values */
+  int NNZ = getNNZ();
+  
+  /* Allocate memory for arrays */
```

Comment:
this should actually be something like "Deallocate memory for arrays if previously allocated" or something similar